New file-partition.md doc describing how to partition files to ensure fast initial blockchain synchronization..

jimhashhq opened the pull request:
After native build from source on Mac OS, my initial attempts to synchronize the blockchain were very very slow.  Upon finding [Issue Sync Taking Too Long](https://github.com/bitcoin/bitcoin/issues/10647), I found discussion by all and comments by @sipa in particular to be very useful, and reorganized $datadir folders on my local macOS build/install and summarized steps taken in file-partition.md doc.  These comments might find their audience more appropriately elsewhere, please feel free to suggest, thank you very much.
-jimhash
Note: This looks to be logged as issue: https://github.com/bitcoin/bitcoin/issues/10736

jimhashhq committed:
Create file-partition.md

Describe partitioning of datadir files between the high-frequency/low-capacity "index" files and the low-frequency/high-capacity "blocks" files.  These steps are probably obvious to more adept bitcoind admins, but for newbies, like myself I didn't see these steps written anywhere else.

jimhashhq committed:
Update file-partition.md

Formatting

jimhashhq committed:
Update file-partition.md

Typos.

jimhashhq committed:
Update file-partition.md

jimhashhq committed:
Update file-partition.md

sipa commented:
I think you missed something. There are three directories that matter:
* $DATADIR/blocks
* $DATADIR/blocks/index
* $DATADIR/chainstate

The first is high-bandwidth, low IO. The second has hardly any activity at all. The third is where all activity happens, and is critical for performance.

You should not separate the blocks from the blocks/index directory, as things may get ugly if they're out of sync. You can however put the chainstate on a faster/smaller device.

jimhashhq commented:
Hi @sipa, thank you very much for the prompt follow up.  Yes, the "chainstate" (in addition to "index") LevelDB files need to stay on the internal disk for reasonable synchronization performance.  The "file-partition.md" notes I made should describe:

- Installing $datadir to local internal disk (includig "chainstate").
- Start/stop bitcoind (though maybe this isn't necessary).
- Move the "index" files up one folder level. 
- Move the "blocks" folder off the internal disk.
- Go to the new external "blocks" folder and soft-link back to the new index location.
- Go to the internal "data" location and link the now external "blocks" folder.

Here, "chainstate" remains on the internal disk as never needs to move.

Again these "file-partition.md" notes are not the direct route I took, but (if correct) might save the next person considerable delay in initial sync.

To your point even though these low-capacity/high-IOfrequency LevelDb and high-capacity/low-IOfrequency blockchain files are physically separated on separate disks, they are kept in sync via the soft links created.

_Maybe ultimately a more natural configuration of these files would be to have the "index" folder up a level to start with._  That's what makes moving these folders confusing in my opinion.

Thank you again very much for the prompt feedback, very flattering, thanks.

jimhashhq committed:
Update file-partition.md

Per @sipa, add mention of the fact that, like the "index" folder, the "chainstate" LevelDB folder must also remain on a fast (internal) drive if reasonable synchronization time is to be expected.

jimhashhq committed:
Update file-partition.md

jimhashhq committed:
Update file-partition.md

jimhashhq commented:
I updated the doc to describe the need to keep "chainstate" on a fast internal disk as per @sipa, thank you.  Also, thanks @fanquake for adding the "Docs & Output" label.  This issue is probably also related to installation and/or configuration, but I did not see a label for that in the bitcoin project, perhaps that might be a useful label category?  Thank you both.

jimhashhq commented:
"file-partition.md"has been updated to highlight the need to keep "chainstate" folder on a fast (internal) drive in addition to "index" files as per @sipa.  Thanks @fanquake for adding the label.   This issue seems also related to "installation" and/or "configuration"; I did not see either of these as bitcoin project label categories, was thinking they might prove useful as well.  Thank you both for your guidance.

achow101 commented on doc/file-partition.md:
```diff
@@ -0,0 +1,53 @@
+
+2) Stop bitcoind, so that we can rearrange some datadir folders:
+
+       kill -QUIT `cat /Volumes/WD-Passport-Mac/bitcoin/data/bitcoind.pid`
```
Instead of killing the process, you should use `bitcoin-cli stop`.

achow101 commented on doc/file-partition.md:
```diff
@@ -0,0 +1,53 @@
+      
+          ln -s /Volumes/WD-Passport-Mac/bitcoin/blocks /Users/coinadm/local/bitcoin/data/blocks
+
+ 注意 - Nota - Note - ध्यान दें - ﻢﻠﺣﻮﻇﺓ - метка 
```
What's with the muptiple languages here?

jimhashhq replied:
Again, good question, thank you.  To me, the "key finding" here (if any, really) is simply that the "index" folder by default is nested *within* the "blocks" folder, unlike the "chainstate" folder which is a sibling folder of "blocks".  This *slightly* complicates moving the "blocks" folder off of the internal disk to an external disk; my apologies for repeating the obvious.  Database administrators out there (me included) might argue that this nested folder configuration is less desirable as it complicates physical separation of high-capacity/low-frequency block files from the lower-capacity/higher-frequency index (and chainstate) LevelDB files.
When I open the hood to my car there are warnings labels on the radiator cap, etc., etc.., and these labels are in multiple languages to point out appropriate cautions to naive vehicle operators who may have never looked under the hood or checked a radiator.  I'm taking inspiration from this and wanted to (hopefully) say "Note" in 1/2 dozen or so most common languages by usage.  Also, I want to be especially friendly in this day and age; I feel like we could all use it.
In closing, I hope it doesn't sound like I am trying to "make a mountain out of a molehill" here, it's not that at all; I just wanted to share my personal experiences in hopes of further facilitating ease of use of the system.  Overall I find the system to be very easy to work with and well thought out.

jimhashhq committed:
Update file-partition.md

Correct the command for step #2 in this outline; per @achow101, thanks very much.

sipa commented:
Your document is still suggesting to split the blocks/ directory from the blocks/index/ directory. Please don't do that; it's dangerous (they need to be in sync), and unnecessary (the blocks/index/ directory hardly sees any I/O). You should just suggest to move the chainstate/ to a faster drive compared to blocks/.

jimhashhq commented on doc/file-partition.md:
```diff
@@ -0,0 +1,53 @@
+| ${datadir}/blocks/index | ${datadir}   | low            | high           |
+| ${datadir}/blocks       | ${EXTERAL}   | high           | low            |
+| ${datadir}/chainstate   | n/a          | low            | high           |
+
```
There seems to be some agreement that the last two rows in the above table are correct.  At issue still is whether the *first row* is correct or not -- it does reflect my experience, and is the impetus for this doc.  I identified this possible re-configuration pitfall (which I mistakenly made myself) via bitcoind file usage reported by the following command:
     lsof -p `cat ${datadir}/bitcoind.pid` | grep ldb\$
This motivated me to move the subordinate/child "index" folder back onto the internal drive and set up the soft links described here.  I did not "go the extra mile" though and verify i/o frequency demands of these index LeveDB files.  If someone could chime in and corroborate either way that would be appreciated.  Is there a blockchain indexing or synchronization test that we can run with the index files both *on* the external drive and *off-of* the external drive while capturing i/o frequency statistics as well as wall clock time?  Arguably this is currently "seat of my pants"; I first moved the entire blocks folder (including the index subfolder) to the external disk for internal capacity reasons, and it synched super super slowly.  Then I looked at open files as described above and spotted those "LevelDB" index folder files.  Moving them as described here sped things up to the point that performance seemed to match that with the default configuration of the blocks folder on the internal drive.  Basically I was just doing a simplistic `ls -lrt blk*.dat` every so often and watching how fast the blk*.dat files were growing in both cases.

A possible misperception here is that this doc is meant to be some sort of performance advice.  It is not, rather it's really just meant to be a note to help other developers/analysts who (like me) want to work on very low end commodity hardware yet perform initial synchronization quickly, without the large-capacity internal disk space requirements.

Neither are these meant as instructions to backup the blockchain for portability between bitcoin development instances either; as @sipa points out, the "blocks" folder 'needs to be kept in sync with the "index" folder', and the blocks folder is useless by itself for sure.  

Missing from this document is the (?reasonable?) expectation that bitcoind not be started until the external disk is mounted and likewise that the external disk not be dismounted/ejected until bitcoind is shut down.   To this point, I have not tried to see what happens, if after running in the configuration suggested here, the operator/developer accidentally tries to start bitcoin w/o the the external storage plugged in; I would hope that the index folder files are not corrupted if the blocks folder is not accessible.

This is merely intended to be more of a "pitfalls to avoid" type of document, rather than some sort of tuning advice.  I hope that this perspective is of use.

flack commented on doc/file-partition.md:
```diff
@@ -0,0 +1,53 @@
+| Folder Name              | Link Name           |
+| ------------------------ | ------------------- |
+| ${EXTERNAL}/blocks/index | ${datadir}/../index |
+| ${datadir}/blocks        | ${EXTERAL}/blocks   |
```
typo: `EXTERAL`

laanwj commented:
> (the blocks/index/ directory hardly sees any I/O). 

Except with `-txindex` I guess :(

laanwj commented:
> Please don't do that; it's dangerous (they need to be in sync)

Also: sometimes there's the problem with the leveldb not supporting the filesystem that the `blocks/` directory points to. I think this happens with some network filesystems. If `blocks/index` is on the same partition that will never work :/

jimhashhq committed:
Update file-partition.md

jimhashhq commented:
My apologies for not getting back sooner, I was ill but am feeling better.
I think my intent here was just to share/communicate my experiences with "symlink(ing) out the block-files (the large part)" as mentioned in issue #10787 referenced above, thanks.
I was hoping this experience might prove useful to others on low end commodity hardware who wish to store the blocks on an external USB.  Thanks!

arowser commented on doc/file-partition.md:
```diff
@@ -0,0 +1,53 @@
+      
+          ln -s /Users/coinadm/local/bitcoin/index /Volumes/WD-Passport-Mac/bitcoin/blocks/index 
+
+6) Replace the original index folder location with a soft link:
```
should be "block folder"?

jimhashhq replied:
Corrected, thanks!

jimhashhq committed:
Update file-partition.md

Correct typo per @arowser, thank you.

laanwj commented:
> I was hoping this experience might prove useful to others on low end commodity hardware who wish to store the blocks on an external USB. Thanks!

Why close?

jimhashhq commented:
I got my GitHub workflow wires crossed, reopening thank you.
